scrapy/middleware.py

```diff
@@ -50,8 +51,41 @@ def __init__(self, *middlewares: Any) -> None:
     def _get_mwlist_from_settings(cls, settings: Settings) -> list[Any]:
         raise NotImplementedError
 
+    @staticmethod
+    def _build_from_settings(objcls: type[_T], settings: BaseSettings) -> _T:
+        if hasattr(objcls, "from_settings"):
+            warnings.warn(
+                f"{objcls.__qualname__} has from_settings() but not from_crawler()."
```

Comment:
I suspect I am missing something, but from a quick look, it seems to be that `_build_from_settings` is called when `MiddlewareManager.from_settings` is used, and I see nothing ensuring that `objcls` does not have `from_crawler` defined.